sympy/assumptions/facts.py

```diff
@@ -34,71 +36,138 @@ def get_composite_predicates():
-        Implies(Q.negative_infinite, ~Q.positive_infinite),
-        Implies(Q.negative, ~Q.zero & ~Q.positive),
-        Implies(Q.positive, ~Q.zero),
+        Implies(Q.negative_infinite(x), ~Q.positive_infinite(x)),
+        Implies(Q.negative(x), ~Q.zero(x) & ~Q.positive(x)),
+        Implies(Q.positive(x), ~Q.zero(x)),
 
         # build real line and complex plane
-        Implies(Q.negative | Q.zero | Q.positive, ~Q.imaginary),
-        Implies(Q.negative | Q.zero | Q.positive | Q.imaginary, Q.algebraic | Q.transcendental),
+        Implies(Q.negative(x) | Q.zero(x) | Q.positive(x), ~Q.imaginary(x)),
+        Implies(Q.negative(x) | Q.zero(x) | Q.positive(x) | Q.imaginary(x),
+                Q.algebraic(x) | Q.transcendental(x)),
 
         # other subsets of complex
-        Implies(Q.transcendental, ~Q.algebraic),
-        Implies(Q.irrational, ~Q.rational),
-        Equivalent(Q.rational | Q.irrational, Q.negative | Q.zero | Q.positive),
-        Implies(Q.rational, Q.algebraic),
+        Implies(Q.transcendental(x), ~Q.algebraic(x)),
+        Implies(Q.irrational(x), ~Q.rational(x)),
+        Equivalent(Q.rational(x) | Q.irrational(x),
+                Q.negative(x) | Q.zero(x) | Q.positive(x)),
+        Implies(Q.rational(x), Q.algebraic(x)),
 
         # integers
-        Implies(Q.even, ~Q.odd),
-        Implies(Q.even | Q.odd, Q.rational),
-        Implies(Q.zero, Q.even),
-        Implies(Q.composite, ~Q.prime),
-        Implies(Q.composite | Q.prime, (Q.even | Q.odd) & Q.positive),
-        Implies(Q.even & Q.positive & ~Q.prime, Q.composite),
+        Implies(Q.even(x), ~Q.odd(x)),
+        Implies(Q.even(x) | Q.odd(x), Q.rational(x)),
+        Implies(Q.zero(x), Q.even(x)),
+        Implies(Q.composite(x), ~Q.prime(x)),
+        Implies(Q.composite(x) | Q.prime(x),
+            (Q.even(x) | Q.odd(x)) & Q.positive(x)),
+        Implies(Q.even(x) & Q.positive(x) & ~Q.prime(x), Q.composite(x)),
 
         # hermitian and antihermitian
-        Implies(Q.negative | Q.zero | Q.positive, Q.hermitian),
-        Implies(Q.imaginary, Q.antihermitian),
-        Implies(Q.zero, Q.hermitian | Q.antihermitian),
+        Implies(Q.negative(x) | Q.zero(x) | Q.positive(x),
+            Q.hermitian(x)),
+        Implies(Q.imaginary(x), Q.antihermitian(x)),
+        Implies(Q.zero(x), Q.hermitian(x) | Q.antihermitian(x)),
 
         # define finity and infinity, and build extended real line
-        Implies(Q.infinite, ~Q.finite),
-        Implies(Q.algebraic | Q.transcendental, Q.finite),
-        Implies(Q.negative_infinite | Q.positive_infinite, Q.infinite),
+        Implies(Q.infinite(x), ~Q.finite(x)),
+        Implies(Q.algebraic(x) | Q.transcendental(x), Q.finite(x)),
+        Implies(Q.negative_infinite(x) | Q.positive_infinite(x), Q.infinite(x)),
 
         # commutativity
-        Implies(Q.finite | Q.infinite, Q.commutative),
+        Implies(Q.finite(x) | Q.infinite(x), Q.commutative(x)),
```

Comment:
This is extremely weird. I think maybe there used to be something like this in the old assumptions but I got rid of it and just made `finite` and `infinite` mutually exclusive. There is a need in the old assumptions for the `is_commutative` attribute although perhaps it should be completely separate from the assumptions system. I don't think `Q.commutative` is needed in the new assumptions. Is there a clear definition of what it is supposed to even mean?

Reply:
This means that every number (extended real numbers, composite numbers and `zoo`) is finite.

`finite` and `infinite` are not always mutually exclusive, since `S.NaN` is not finite nor infinite.

I'm +1 that we do not need `Q.commutative` in the new assumptions. IMO commutativity must be dealt by kind system and shouldn't be included in old/new assumption system. Removing it from old assumptions is impossible since doing so will raise infinite backwards compatibility issues. But I don't think `Q.commutative` is really used anywhere. Perhaps we can just remove it?

Reply:
Commutative could be removed from the old assumptions in the sense that it can be an attribute/property but just not part of the assumptions system itself. I'm not sure if there is any significant benefit in changing that though.

> `finite` and `infinite` are not always mutually exclusive, since `S.NaN` is not finite nor infinite.

I think NaN can give None for both cases. In all cases where finite=True we should have infinite=False and vice versa so they are mutually exclusive. This particular rule is I think copied from the old assumptions but I have since fixed this in the old assumptions:
https://github.com/sympy/sympy/blob/49df83fdcc04715525b635bb9157113b94d11122/sympy/core/assumptions.py#L257

Reply:
`finite` and `infinite` are now mutually exclusive. What should be done with `Q.commutative`, then?

Reply:
I'm not sure about `Q.commutative` so I would suggest just leaving it for now while working on the refactoring.